[DNM]: Support database read-only (#22266)
pingcap/docs: database-read-only.md

# Database Read-Only

> Database-level read-only allows you to set a single database to read-only, so that write operations (DDL, DML, and some locking reads) against that database and its objects are rejected. This capability is useful in scenarios such as sharded migration, traffic failback drills, and tenant data export, where you want to reduce the blast radius.

Reviewer: ```suggestion
Database-level read-only allows you to set a single database to read-only, so that write operations (DDL, DML, and some locking reads) against that database and its objects are rejected. 
```

- `DEFAULT` is equivalent to `0` (read-write).
- `READ ONLY` is only supported in `ALTER DATABASE/SCHEMA`, not supported in `CREATE DATABASE`.
- `READ ONLY` must be used alone. It cannot be combined in the same `ALTER DATABASE` statement with other options such as `CHARACTER SET`, `COLLATE`, `PLACEMENT POLICY`, or `SET TIFLASH REPLICA`.

**Examples**

Set a database to read-only:

```sql
ALTER DATABASE test READ ONLY = 1;
```

Set a database back to read-write:

```sql
ALTER DATABASE test READ ONLY = 0;
```

If the database is already in the requested state, the statement succeeds and returns a `Note` (you can check it via `SHOW WARNINGS`).

## Privileges and usage constraints

- **Required privilege**: You must have the `ALTER` privilege on the target database.
- **Interaction with cluster read-only**: When the cluster is in read-only mode (for example, `tidb_restricted_read_only=1`), you cannot change database read-only status. The statement will fail with `Running in read-only mode`.
- **System database restriction**: Database-level read-only is not supported for system databases, including `mysql`, `sys`, `INFORMATION_SCHEMA`, `PERFORMANCE_SCHEMA`, and `METRICS_SCHEMA`.

> 

Reviewer: ```suggestion
- **Bypass privilege**: If a session is explicitly granted the dynamic privilege `RESTRICTED_REPLICA_WRITER_ADMIN`, it can bypass database-level read-only restrictions. This is intended for replica writer / replication-related sessions.
```

```

## Behavior of read-only DDL

### Transaction when switching read-only state

`ALTER DATABASE ... READ ONLY = 1` is an online DDL. It first blocks new writes to the target database and then waits for existing transactions related to that database to commit before returning.

If a transaction has not accessed the database at the beginning, but accesses an object in that database after the read-only DDL finishes, it may fail with an error similar to:

- `public schema <db> read only state has changed`

### Troubleshooting when DDL takes a long time

If `ALTER DATABASE ... READ ONLY = 1` takes a long time to finish, it is usually blocked by uncommitted transactions, TiDB periodically outputs the following logs:

```
[INFO] [schema.go:xxx] ["uncommitted txn block read only ddl"] [category=ddl] ["txn ID"=<txn_id>] ["job ID"=<job_id>]
```

You can query transaction information and locate the blocked session to further confirm the scope of the impact:

```sql
SELECT INSTANCE, ID, SESSION_ID, USER, DB, STATE, START_TIME, RELATED_TABLE_IDS FROM INFORMATION_SCHEMA.CLUSTER_TIDB_TRX;
```

## Behavior in a read-only database

> When a database is read-only, restricted statements typically fail with:

Reviewer: ```suggestion
Unless the session has the `RESTRICTED_REPLICA_WRITER_ADMIN` dynamic privilege, restricted statements typically fail with:
```
